### DISCUSSION TITLE
conditional deprecation

### DISCUSSION THREAD

**Discussion by dchelimsky** [clj-kondo/clj-kondo#1982] [Q&A]
To upvote this issue, give it a thumbs up. See [this list](https://github.com/clj-kondo/clj-kondo/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) for the most upvoted issues.

**Is your feature request related to a problem? Please describe.**

We have a utility function library with `map-vals`, that does the same thing as clojure-1.11's `clojure.core/update-vals`, but predates it. We have some consumers that are using clojure-1.11, but some that are using older versions. We'd like to convey to consumers using 1.11 that our home-grown `map-vals` is deprecated and you should use `clojure.core/update-vals`, but consumers using < 1.11 would not see this.

**Describe the solution you'd like**

Here's one idea, but I'm sure there are plenty of others that would work: add support for e.g. `:deprecated {:doc "..." :replaced-by {:lib "org.clojure/clojure" :version "1.11.0"}}`. This would support the idea for any library at all.

`clj-kondo` would have to look validate this against the project's dependencies.

**Describe alternatives you've considered**

The keys could be named differently, and/or could be flattened (all at top level of the metadata map)

**Comment by borkdude** [comment]
I'm converting this to a Q&A since I'm fairly sure clj-kondo already has what you need. The linter you're searching for is probably:

https://github.com/clj-kondo/clj-kondo/blob/master/doc/linters.md#discouraged-var

You can make a configuration for all of your libs which discourages `your.lib/map-vals` with a message.

Projects that still use clojure 1.11 can override this config in their local `.clj-kondo/config.edn`.

**Reply by borkdude** [reply]
I'd have to double check if you can undo discouraging in a local config. There might be an issue with that, see:

https://github.com/clj-kondo/clj-kondo/issues/1979

So it would be good to watch that issue.
